addOns/paramdigger/src/main/java/org/zaproxy/addon/paramdigger/CacheController.java

```diff
@@ -0,0 +1,864 @@
+                    msg.setRequestHeader(headers);
+
+                    httpSender.sendAndReceive(msg);
+                    times.add(msg.getTimeElapsedMillis());
+
+                    if (msg.getResponseHeader().getStatusCode()
+                            != base.getResponseHeader().getStatusCode()) {
+                        // TODO show output that unexpected response code was received
+                    }
+                }
+
+                boolean skip = false;
+                for (int j = 1; j < times.size(); j++) {
+                    if ((j % 2 == 1)
+                            && (times.get(j - 1) - times.get(j)
+                                    < config.getCacheBustingThreshold())) {
+                        /* Since the response was faster then usual timing. We can assume it came from a cache. */
+                        skip = true;
+                        break;
+                    }
+                }
+                if (skip) {
+                    continue;
+                }
+                /* There is a cache and cache buster works! */
+                cache.setTimeIndicator(true);
+                cache.setCacheBusterFound(true);
+                cache.setCacheBusterIsHttpMethod(true);
+                cache.setCacheBusterName(httpMethods[i]);
+                this.bustedMessage = msg;
+            }
+        }
+    }
+
+    /**
+     * Checks if a cookie can be used as a cache buster. Requires users to specify what would be the
+     * list of cookies that can be tried as cache busters.
+     *
+     * @param url The URL to check.
+     * @param method The HTTP Method to check.
+     * @param cache The cache object to store the results in.
+     * @throws NullPointerException if the URL is null or empty
+     * @throws IOException if request was not sent.
+     * @throws IllegalArgumentException if the HTTP Method is not supported.
+     */
+    private void cacheBusterCookie(String url, Method method, Cache cache)
+            throws NullPointerException, IOException, IllegalArgumentException {
+        List<String> cookies = config.getCacheBustingCookies();
+        for (int i = 0; i < cookies.size(); i++) {
+            HttpRequestHeader headers = new HttpRequestHeader();
+            headers.setURI(new URI(url, true));
+            headers.setVersion(HttpHeader.HTTP11);
+            switch (method) {
+                case GET:
+                    headers.setMethod(HttpRequestHeader.GET);
+                    break;
+                case POST:
+                    headers.setMethod(HttpRequestHeader.POST);
+                    break;
+                default:
+                    throw new IllegalArgumentException(
+                            Constant.messages.getString(METHOD_NOT_SUPPORTED, method));
+            }
+            if (cache.getIndicator() != null && !cache.getIndicator().isEmpty()) {
+                String cb = (Integer.valueOf(random.nextInt(RANDOM_SEED))).toString();
```

Comment:
I know you're still working on this, but making an `Integer` just to toString() it seems inefficient. Couldn't this be simplified to something like?
```suggestion
                String cb = String.valueOf(random.nextInt(RANDOM_SEED));
```

Reply:
Yep :)

Reply:
The major problem is still the timing tests

Reply:
@kingthorin can you pull this and try it out in your local ... I can't find out what or why I'm missing the tests :)

Reply:
Sure, but I won't get to it for a few hours.

Reply:
Uhh the quicker the better :)... I can only please in a cute voice :")

Reply:
Thread.sleep is in [millis](https://docs.oracle.com/javase/7/docs/api/java/lang/Thread.html#sleep(long)) so 60 vs 70 is essentially imperceptible. I know we're talking computers and timing, and I haven't futzed with this at all.

Reply:
Uhhh :) then how to decide caching based on timing ?? :)

Reply:
Uhhh :) then how to decide caching based on timing ?? :)

Reply:
the average delay is just 20 ms

Reply:
cached responses take 5-6 ms whereas direct from server can take a minimum kinda 20 ms

Reply:
Okay, that should be workable here then. I guess you could actually time things. Don't worry about it for now, let me pull it and see what I can sort out.

Reply:
Have you been able to try it against something "live" even if it's a vuln app?

Reply:
Uhh no. How can try this?

Reply:
I can try it on any website .. but how to do it?

Reply:
You'd have to find a vuln app that manifests the thing you're trying to test for. I thought you might have queued some up when prepping for the project. You could try searching at bit now but I would not recommend getting stuck hunting.

Reply:
I do have a bunch of them. But they don't exhibit the timing based caching mechanism

Reply:
Okay, don't worry about it then.

Reply:
No problem. Doing things based on timing has proven challenging elsewhere (SQLi scan rules etc) as targets is very variable response wise (based on other users/activity, based on ZAP being multi-threaded, etc. we've found it's pretty easy to disrupt things timing wise [intentionally or not]).

Reply:
The static tests pass smoothly :).

Reply:
You could add `@Disable` (or whatever the annotation is) to the particular unit test method and add a comment that it requires further work/investigation. (For my two cents anyway.)

Reply:
At this point .... I feel that would be really better :). I have this weird problem that if I cannot find the fault in 5 minutes I cannot find it at all :). For this one I noticed the problem a lot later... And the solution hasn't arrived in my brain ... Probably it will never arrive :)

Reply:
You may have also been staring at it too long at this point. Just move along to something else.